File: typescript/src/hosting/warmup.ts
// hosting/warmup.ts
//
// Port of the CircleAI.Hosting.Warmup surface (RT-07):
//   • IRequestPredictor.cs / ArrivalForecast
//   • HistogramRequestPredictor.cs — per-minute-of-day EWMA arrival histogram
//   • PredictiveWarmupOptions.cs
//   • PredictiveWarmupController.cs — polls the predictor, pre-warms on a spike
//
// All arithmetic is C# `double`; there are no `float` sites, so no Math.fround
// is required. The controller's background loop is driven by setInterval; a
// single-cycle tickAsync is exposed for tests.

import type { IAIService } from "./service.js";

/**
 * Forecast of inbound requests over a window. Mirrors
 * CircleAI.Hosting.Warmup.ArrivalForecast.
 */
export interface ArrivalForecast {
  /** 0.0 .. 1.0 — chance of at least one arrival in the window. */
  readonly probabilityOfArrival: number;
  /** Best estimate of how many arrivals to expect. */
  readonly expectedCount: number;
  /** 0.0 .. 1.0 — trustworthiness given sample size; ~0 at cold-start. */
  readonly confidence: number;
}

/**
 * Local-only predictor that learns request arrival timing. Mirrors
 * CircleAI.Hosting.Warmup.IRequestPredictor.
 */
export interface IRequestPredictor {
  /** Record one arrival at `utc`. */
  recordArrival(utc: Date): void;
  /** Forecast arrivals over `forecastWindowMs` starting at `utcNow`. */
  predict(utcNow: Date, forecastWindowMs: number): ArrivalForecast;
  /** Total arrivals observed since construction. */
  readonly observedArrivals: number;
}

const MINUTES_PER_DAY = 24 * 60;
const WARM_CONFIDENCE = 1.0;
const MIN_SAMPLES_FOR_FULL_CONFIDENCE = 25;

/**
 * Default {@link IRequestPredictor} — a rolling per-minute-of-day EWMA arrival
 * histogram. Mirrors CircleAI.Hosting.Warmup.HistogramRequestPredictor.
 */
export class HistogramRequestPredictor implements IRequestPredictor {
  private readonly historyDays: number;
  private readonly perMinuteRate: Float64Array;
  private readonly perMinuteCount: Int32Array;
  private observed = 0;

  constructor(historyDays = 7) {
    if (historyDays <= 0) throw new Error("historyDays must be positive.");
    this.historyDays = historyDays;
    this.perMinuteRate = new Float64Array(MINUTES_PER_DAY);
    this.perMinuteCount = new Int32Array(MINUTES_PER_DAY);
  }

  get observedArrivals(): number {
    return this.observed;
  }

  recordArrival(utc: Date): void {
    const minute = utc.getUTCHours() * 60 + utc.getUTCMinutes();
    const cnt = ++this.perMinuteCount[minute];
    // EWMA over the last `historyDays` observations at this slot.
    const alpha = 2.0 / (Math.min(cnt, this.historyDays) + 1);
    this.perMinuteRate[minute] =
      alpha * 1.0 + (1 - alpha) * this.perMinuteRate[minute];
    this.observed++;
  }

  predict(utcNow: Date, forecastWindowMs: number): ArrivalForecast {
    if (forecastWindowMs <= 0)
      return { probabilityOfArrival: 0, expectedCount: 0, confidence: 0 };
    if (this.observed === 0)
      return { probabilityOfArrival: 0, expectedCount: 0, confidence: 0 };

    const minute = utcNow.getUTCHours() * 60 + utcNow.getUTCMinutes();
    const minutes = Math.max(1, Math.ceil(forecastWindowMs / 60000));
    let expected = 0;
    let coveredSamples = 0;
    for (let i = 0; i < minutes; i++) {
      const idx = (minute + i) % MINUTES_PER_DAY;
      expected += this.perMinuteRate[idx];
      coveredSamples += this.perMinuteCount[idx];
    }

    // Poisson tail: P(>=1 arrival) = 1 - exp(-lambda).
    const probability = 1.0 - Math.exp(-expected);
    const confidence = Math.min(
      WARM_CONFIDENCE,
      coveredSamples / (MIN_SAMPLES_FOR_FULL_CONFIDENCE * minutes),
    );
    return {
      probabilityOfArrival: probability,
      expectedCount: expected,
      confidence,
    };
  }

  /** Test-only — wipe state. */
  resetForTests(): void {
    this.perMinuteRate.fill(0);
    this.perMinuteCount.fill(0);
    this.observed = 0;
  }
}

/**
 * Configuration for {@link PredictiveWarmupController}. Mirrors
 * CircleAI.Hosting.Warmup.PredictiveWarmupOptions. Durations are milliseconds.
 */
export interface PredictiveWarmupOptions {
  /** Opt-in to pre-warming. Default false. */
  readonly enabled?: boolean;
  /** How often to poll the predictor, ms. Default 30 000. */
  readonly pollIntervalMs?: number;
  /** How far ahead to forecast, ms. Default 60 000. */
  readonly forecastWindowMs?: number;
  /** Pre-warm when probability × confidence ≥ this. Default 0.5. */
  readonly warmupThreshold?: number;
  /** Minimum delay between consecutive pre-warm calls, ms. Default 300 000. */
  readonly minTimeBetweenWarmupsMs?: number;
}

/** Resolved defaults for {@link PredictiveWarmupOptions}. */
export const DEFAULT_PREDICTIVE_WARMUP_OPTIONS: Required<PredictiveWarmupOptions> = {
  enabled: false,
  pollIntervalMs: 30_000,
  forecastWindowMs: 60_000,
  warmupThreshold: 0.5,
  minTimeBetweenWarmupsMs: 5 * 60_000,
};

/**
 * Background loop that polls an {@link IRequestPredictor} and triggers
 * {@link IAIService.prewarmAsync} before predicted spikes. Mirrors
 * CircleAI.Hosting.Warmup.PredictiveWarmupController.
 */
export class PredictiveWarmupController {
  private readonly service: IAIService;
  private readonly predictor: IRequestPredictor;
  private readonly options: Required<PredictiveWarmupOptions>;
  private readonly clock: () => Date;

  private timer: ReturnType<typeof setInterval> | null = null;
  private started = false;
  private disposed = false;
  private lastWarmupMs = Number.NEGATIVE_INFINITY;

  constructor(
    service: IAIService,
    predictor: IRequestPredictor,
    options: PredictiveWarmupOptions = {},
    clock?: () => Date,
  ) {
    if (!service) throw new Error("service required");
    if (!predictor) throw new Error("predictor required");
    this.service = service;
    this.predictor = predictor;
    this.options = { ...DEFAULT_PREDICTIVE_WARMUP_OPTIONS, ...options };
    this.clock = clock ?? (() => new Date());
  }

  /** Begin polling. No-op when disabled or already started. */
  async startAsync(): Promise<void> {
    if (this.disposed) throw new Error("PredictiveWarmupController is disposed.");
    if (!this.options.enabled || this.started) return;
    this.started = true;
    // Run one tick immediately (mirrors the C# do/while loop).
    await this.tickAsync();
    this.timer = setInterval(() => {
      void this.tickAsync();
    }, this.options.pollIntervalMs);
    (this.timer as { unref?: () => void }).unref?.();
  }

  /** Record a request arrival on the predictor at "now". */
  notifyArrival(): void {
    this.predictor.recordArrival(this.clock());
  }

  /**
   * One prediction + decide-and-maybe-warm cycle. Returns true when warmup
   * fired. Public for tests and manual poking.
   */
  async tickAsync(): Promise<boolean> {
    const now = this.clock();
    const forecast = this.predictor.predict(now, this.options.forecastWindowMs);
    const score = forecast.probabilityOfArrival * forecast.confidence;
    if (score < this.options.warmupThreshold) return false;
    if (now.getTime() - this.lastWarmupMs < this.options.minTimeBetweenWarmupsMs)
      return false;

    try {
      this.lastWarmupMs = now.getTime();
      await this.service.prewarmAsync();
      return true;
    } catch {
      return false;
    }
  }

  async disposeAsync(): Promise<void> {
    if (this.disposed) return;
    this.disposed = true;
    if (this.timer !== null) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }
}

File: rust/src/hosting/warmup.rs
//! warmup.rs
//!
//! (RT-07) Local-only request-timeline learner + predictive warmup controller.
//! Ported from `Warmup/IRequestPredictor.cs`, `Warmup/HistogramRequestPredictor.cs`,
//! `Warmup/PredictiveWarmupOptions.cs`, and `Warmup/PredictiveWarmupController.cs`.
//!
//! The predictor records arrival times and forecasts whether a spike is coming;
//! the controller polls it and pre-warms the generator before a predicted spike.
//! All counting is in-process — no telemetry, no upload. The C# controller runs
//! an async `PeriodicTimer` loop; the sync port exposes
//! [`PredictiveWarmupController::tick`] which the host calls on its own interval.

use std::sync::Mutex;

use chrono::{DateTime, Duration, Timelike, Utc};

use super::service::IAIService;

/// (RT-07) Forecast of inbound requests over a window. 1:1 with the C#
/// `ArrivalForecast` readonly record struct.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ArrivalForecast {
    /// 0.0 .. 1.0 — probability of at least one arrival in the window.
    pub probability_of_arrival: f64,
    /// Best estimate of how many arrivals to expect.
    pub expected_count: f64,
    /// 0.0 .. 1.0 — how trustworthy the estimate is given sample size.
    pub confidence: f64,
}

impl ArrivalForecast {
    pub fn new(probability_of_arrival: f64, expected_count: f64, confidence: f64) -> Self {
        Self {
            probability_of_arrival,
            expected_count,
            confidence,
        }
    }
}

/// (RT-07) Local-only predictor that learns request arrival timing. 1:1 with the
/// C# `IRequestPredictor`.
pub trait IRequestPredictor: Send + Sync {
    /// Record one arrival at `utc`.
    fn record_arrival(&self, utc: DateTime<Utc>);

    /// Forecast arrivals in `forecast_window` starting at `utc_now`.
    fn predict(&self, utc_now: DateTime<Utc>, forecast_window: Duration) -> ArrivalForecast;

    /// Total arrivals observed since construction.
    fn observed_arrivals(&self) -> i64;
}

const MINUTES_PER_DAY: usize = 24 * 60;
const WARM_CONFIDENCE: f64 = 1.0;
const MIN_SAMPLES_FOR_FULL_CONFIDENCE: i64 = 25;

struct HistogramInner {
    /// index = minute-of-day; value = avg arrivals/minute observed (EWMA).
    per_minute_rate: Vec<f64>,
    per_minute_count: Vec<i32>,
}

/// (RT-07) Default [`IRequestPredictor`] — keeps a histogram of per-minute
/// arrival rates over a rolling window of recent days. 1:1 with the C#
/// `HistogramRequestPredictor` (same EWMA, same Poisson-tail probability, same
/// confidence formula).
pub struct HistogramRequestPredictor {
    history_days: i64,
    inner: Mutex<HistogramInner>,
    observed: Mutex<i64>,
}

impl HistogramRequestPredictor {
    /// Construct with a rolling history of `history_days` days. Default 7.
    /// Panics on `history_days <= 0` (matches the C#
    /// `ArgumentOutOfRangeException`).
    pub fn new(history_days: i64) -> Self {
        assert!(history_days > 0, "historyDays must be positive");
        Self {
            history_days,
            inner: Mutex::new(HistogramInner {
                per_minute_rate: vec![0.0; MINUTES_PER_DAY],
                per_minute_count: vec![0; MINUTES_PER_DAY],
            }),
            observed: Mutex::new(0),
        }
    }

    /// Default 7-day history (one calendar week).
    pub fn with_default_history() -> Self {
        Self::new(7)
    }

    /// Test-only — wipe state. 1:1 with the C# `ResetForTests`.
    pub fn reset_for_tests(&self) {
        let mut inner = self.inner.lock().unwrap();
        inner.per_minute_rate.iter_mut().for_each(|r| *r = 0.0);
        inner.per_minute_count.iter_mut().for_each(|c| *c = 0);
        *self.observed.lock().unwrap() = 0;
    }
}

impl IRequestPredictor for HistogramRequestPredictor {
    fn record_arrival(&self, utc: DateTime<Utc>) {
        let minute = (utc.hour() as usize * 60) + utc.minute() as usize;
        {
            let mut inner = self.inner.lock().unwrap();
            inner.per_minute_count[minute] += 1;
            let cnt = inner.per_minute_count[minute] as i64;
            // EWMA over the last `history_days` of observations at this slot.
            let alpha = 2.0 / (cnt.min(self.history_days) as f64 + 1.0);
            inner.per_minute_rate[minute] =
                (alpha * 1.0) + ((1.0 - alpha) * inner.per_minute_rate[minute]);
        }
        *self.observed.lock().unwrap() += 1;
    }

    fn predict(&self, utc_now: DateTime<Utc>, forecast_window: Duration) -> ArrivalForecast {
        if forecast_window <= Duration::zero() {
            return ArrivalForecast::new(0.0, 0.0, 0.0);
        }
        let observed = self.observed_arrivals();
        if observed == 0 {
            return ArrivalForecast::new(0.0, 0.0, 0.0);
        }

        let minute = (utc_now.hour() as usize * 60) + utc_now.minute() as usize;
        // Ceiling of window in minutes, at least 1.
        let window_secs = forecast_window.num_seconds() as f64;
        let minutes = ((window_secs / 60.0).ceil() as i64).max(1) as usize;

        let mut expected = 0.0_f64;
        let mut covered_samples = 0_i64;
        {
            let inner = self.inner.lock().unwrap();
            for i in 0..minutes {
                let idx = (minute + i) % MINUTES_PER_DAY;
                expected += inner.per_minute_rate[idx];
                covered_samples += inner.per_minute_count[idx] as i64;
            }
        }

        // Poisson tail: P(>=1 arrival) = 1 - exp(-lambda).
        let probability = 1.0 - (-expected).exp();
        let confidence = WARM_CONFIDENCE.min(
            covered_samples as f64 / (MIN_SAMPLES_FOR_FULL_CONFIDENCE * minutes as i64) as f64,
        );
        ArrivalForecast::new(probability, expected, confidence)
    }

    fn observed_arrivals(&self) -> i64 {
        *self.observed.lock().unwrap()
    }
}

/// (RT-07) Configuration for [`PredictiveWarmupController`]. 1:1 with the C#
/// `PredictiveWarmupOptions`.
#[derive(Debug, Clone, Copy)]
pub struct PredictiveWarmupOptions {
    /// When `false` (default), the controller does not pre-warm.
    pub enabled: bool,
    /// How often the controller polls the predictor. Default 30 s.
    pub poll_interval: Duration,
    /// How far ahead to forecast. Default 60 s.
    pub forecast_window: Duration,
    /// Pre-warm when `probability × confidence >= threshold`. Default 0.5.
    pub warmup_threshold: f64,
    /// Minimum delay between consecutive pre-warm calls. Default 5 minutes.
    pub min_time_between_warmups: Duration,
}

impl Default for PredictiveWarmupOptions {
    fn default() -> Self {
        Self {
            enabled: false,
            poll_interval: Duration::seconds(30),
            forecast_window: Duration::seconds(60),
            warmup_threshold: 0.5,
            min_time_between_warmups: Duration::minutes(5),
        }
    }
}

/// A monotonic clock provider (mirrors the C# `Func<DateTimeOffset>`).
pub type Clock = Box<dyn Fn() -> DateTime<Utc> + Send + Sync>;

/// (RT-07) Polls an [`IRequestPredictor`] and triggers [`IAIService`] pre-warm
/// before predicted spikes. 1:1 with the C# `PredictiveWarmupController` (the
/// async `PeriodicTimer` loop is replaced by the host calling [`Self::tick`]).
pub struct PredictiveWarmupController<'a, P: IRequestPredictor> {
    service: &'a dyn IAIService,
    predictor: &'a P,
    options: PredictiveWarmupOptions,
    clock: Clock,
    last_warmup: Mutex<DateTime<Utc>>,
}

impl<'a, P: IRequestPredictor> PredictiveWarmupController<'a, P> {
    /// Constructs the controller. `clock` defaults to `Utc::now` when `None`.
    pub fn new(
        service: &'a dyn IAIService,
        predictor: &'a P,
        options: PredictiveWarmupOptions,
        clock: Option<Clock>,
    ) -> Self {
        Self {
            service,
            predictor,
            options,
            clock: clock.unwrap_or_else(|| Box::new(Utc::now)),
            last_warmup: Mutex::new(DateTime::<Utc>::MIN_UTC),
        }
    }

    /// Record a request arrival on the underlying predictor at "now". 1:1 with
    /// the C# `NotifyArrival`.
    pub fn notify_arrival(&self) {
        self.predictor.record_arrival((self.clock)());
    }

    /// Run one prediction + decide-and-maybe-warm cycle. Returns `true` when
    /// warmup was triggered. 1:1 with the C# `TickAsync`.
    pub fn tick(&self) -> bool {
        let now = (self.clock)();
        let forecast = self.predictor.predict(now, self.options.forecast_window);
        let score = forecast.probability_of_arrival * forecast.confidence;
        if score < self.options.warmup_threshold {
            return false;
        }
        {
            let last = *self.last_warmup.lock().unwrap();
            if now - last < self.options.min_time_between_warmups {
                return false;
            }
        }

        *self.last_warmup.lock().unwrap() = now;
        match self.service.prewarm() {
            Ok(()) => true,
            Err(_) => false,
        }
    }

    /// Whether the controller is enabled (host loop should skip ticking when
    /// false). Mirrors the C# `StartAsync` early-out on `!Enabled`.
    pub fn is_enabled(&self) -> bool {
        self.options.enabled
    }
}
